src/app/routes/getInitialData/index.test.js

```diff
@@ -0,0 +1,303 @@
+import loggerMock from '#testHelpers/loggerMock'; // Must be imported before getInitialData
+import preprocess from '#lib/utilities/preprocessor';
+import getPreprocessorRules from './utils/getPreprocessorRules';
+
+jest.mock('#lib/utilities/preprocessor', () => jest.fn());
+preprocess.mockImplementation(data => data);
+
+const fetchData = require('./index').default;
+
+describe('fetchData', () => {
+  const mockSuccessfulResponse = {
+    metadata: {},
+    content: {},
+    promo: {},
+  };
+
+  const mockFetchSuccess = () =>
+    fetch.mockResponseOnce(JSON.stringify(mockSuccessfulResponse));
+
+  const mockFetchSuccessWithData = data =>
+    fetch.mockResponseOnce(JSON.stringify(data));
+
+  const mockFetchFailure = () => fetch.mockReject(true);
+
+  const mockFetchInvalidJSON = () => fetch.mockReject('Some Invalid: { JSON');
+
+  const mockFetchNotFoundStatus = () =>
+    fetch.mockResponseOnce(JSON.stringify({}), { status: 404 });
+
+  const mockFetchTeapotStatus = () =>
+    fetch.mockResponseOnce(JSON.stringify({}), { status: 418 });
+
+  const expectedBaseUrl = 'http://localhost';
+  const requestedPathname = '/path/to/asset';
+  const expectedUrl = `${expectedBaseUrl}${requestedPathname}.json`;
+
+  const callfetchData = ({ pathname = requestedPathname, mockFetch }) => {
+    if (mockFetch) {
+      mockFetch();
+    } else {
+      mockFetchSuccess();
+    }
+
+    return fetchData(pathname);
+  };
+
+  afterEach(() => {
+    fetch.resetMocks();
+    jest.clearAllMocks();
+  });
+
+  describe('Succesful fetch', () => {
+    it('should call fetch with correct url', async () => {
+      await callfetchData({});
+
+      expect(fetch).toHaveBeenCalledWith(expectedUrl);
+    });
+
+    it('should call fetch on amp pages without .amp in pathname', async () => {
+      await callfetchData({ pathname: `${requestedPathname}.amp` });
+
+      expect(fetch).toHaveBeenCalledWith(expectedUrl);
+    });
+
+    it('should return an empty object', async () => {
+      const response = await callfetchData({});
+
+      expect(preprocess).toHaveBeenCalledWith(response.pageData, []);
+
+      expect(response).toEqual({
+        pageData: mockSuccessfulResponse,
+        status: 200,
+      });
+    });
+
+    it('should pass preprocessorRules for type: article', async () => {
+      const mockData = {
+        metadata: {
+          type: 'article',
+        },
+      };
+      const expectedPreprocessorRules = getPreprocessorRules('article');
```

Comment:
Can you add a test for `undefined` and `foobar` please? not testing the default case

Reply:
`foorbar` will do. For `undefined` that is covered in the pre-existing tests because the  deafult mock response has metadata as `{}`, can do it explicitly if you want?

Reply:
yeahh an explicit one wont hurt :P